pages/blog/data-cleaning-null-aggregate-values.mdx

```diff
@@ -0,0 +1,354 @@
+  </tr>
+  <tr>
+    <td style={{ border: "1px solid black", padding: "4px" }}>3</td>
+    <td style={{ border: "1px solid black", padding: "4px" }}>42.97</td>
+  </tr>
+  <tr>
+    <td style={{ border: "1px solid black", padding: "4px" }}>4</td>
+    <td style={{ border: "1px solid black", padding: "4px" }}>96.41</td>
+  </tr>
+  <tr>
+    <td style={{ border: "1px solid black", padding: "4px" }}>5</td>
+    <td style={{ border: "1px solid black", padding: "4px" }}>0</td>
+  </tr>
+</table>
+<br />
+
+It works! Fantastic.
+
+<Alert variant="primary">
+  {" "}
+  💡 If all of your foreign keys are coming from a single table, the above might
+  be the right solution for you. However, in more complex cases, you may not have
+  something like a generic “users” table.
+</Alert>
+
+Your primary data may be spread across multiple tables. Fear not, we can still make it happen. Though, for everyone’s sake, at this point I would recommend using a function over using SQL! Just for giggles, let's prove it's possible first.
+
+#### `UNION` with a list
+
+<Alert variant="warning">
+  {" "}
+  ⚠️ Overkill query ahead. You should really try a function instead.
+</Alert>
+
+One way around the issue of having multiple users tables would be to create a list of all the `user_id`s for the query with a sublist of (`user_id`, 0), then `UNION` the original query with that list. If it sounds overkill, that’s because it is:
+
+```sql SELECT DISTINCT on (__pk) __pk, purchases FROM( SELECT user_id as __pk, COUNT(*)
+as purchases FROM demo.purchases WHERE demo.purchases.user_id in (1, 2, 3, 4, 5)
+GROUP BY __pk UNION SELECT * FROM (values (1, 0), (2, 0), (3, 0), (4, 0), (5, 0))
+AS t (__pk, purchases) ORDER BY __pk, purchases DESC ) AS data;
+
+```
+
+Which returns our beloved 0:
+
+<table style={{ textAlign: "center" }}>
+  <thead>
+    <tr>
+      <th style={{ border: "1px solid black" }}> __pk </th>
+      <th style={{ border: "1px solid black" }}> __result </th>
+    </tr>
+  </thead>
+  <tr>
+    <td style={{ border: "1px solid black" }}>1</td>
+    <td style={{ border: "1px solid black" }}>112.50</td>
+  </tr>
+  <tr>
+    <td style={{ border: "1px solid black" }}>2</td>
+    <td style={{ border: "1px solid black" }}>16.48</td>
+  </tr>
+  <tr>
+    <td style={{ border: "1px solid black" }}>3</td>
+    <td style={{ border: "1px solid black" }}>42.97</td>
+  </tr>
+  <tr>
+    <td style={{ border: "1px solid black" }}>4</td>
+    <td style={{ border: "1px solid black" }}>96.41</td>
+  </tr>
+  <tr>
+    <td style={{ border: "1px solid black" }}>5</td>
+    <td style={{ border: "1px solid black" }}>0</td>
+  </tr>
+</table>
+<br />
+
+The query isn’t too out of hand for 5 users, but imagine having to write that out (or have a function write that out) for 500 users! What about 10,000! Just because something works, doesn’t make it necessarily the right idea. So let’s try it from the programming sde.
+
+### Option 2: Using a function 🥇
```

Comment:
I think this title might be misleading for data folks - there /are/ [functions in postgres](https://www.postgresql.org/docs/12/functions.html).  I think what you are talking about is maybe called "Using Code" or "Cleaning Data in Code" maybe?

Reply:
Right.  I mean... SQL is code too.  Hmm.  "After the query" feels most accurate?